cinder/scheduler/filters/sap_affinity_filter.py

```diff
@@ -0,0 +1,199 @@
+LOG = logging.getLogger(__name__)
+
+
+class SAPBackendFilter(filters.BaseBackendFilter):
+
+    run_filter_once_per_request = False
+
+    def __init__(self):
+        super().__init__()
+        self.volume_api = volume.API()
+        self.backend_fqdn_lookup = {}
+
+    def _get_volumes(self, context, affinity_uuids, backend_state):
+        # We don't filter here on host.
+        filters = {'id': affinity_uuids, 'deleted': False}
+        return self.volume_api.get_all(context, filters=filters)
+
+    def _affinity_volumes(self, backend_state, filter_properties):
+        context = filter_properties['context']
+        scheduler_hints = filter_properties.get('scheduler_hints') or {}
+        affinity_uuids = scheduler_hints.get(self.hint_key, [])
+
+        # scheduler hint verification: affinity_uuids can be a list of uuids
+        # or single uuid.  The checks here is to make sure every single string
+        # in the list looks like a uuid, otherwise, this filter will fail to
+        # pass.  Note that the filter does *NOT* ignore string doesn't look
+        # like a uuid, it is better to fail the request than serving it wrong.
+        if isinstance(affinity_uuids, list):
+            for uuid in affinity_uuids:
+                if uuidutils.is_uuid_like(uuid):
+                    continue
+                else:
+                    return False
+        elif uuidutils.is_uuid_like(affinity_uuids):
+            affinity_uuids = [affinity_uuids]
+        else:
+            # Not a list, not a string looks like uuid, don't pass it
+            # to DB for query to avoid potential risk.
+            return False
+
+        if affinity_uuids:
+            return self._get_volumes(context, affinity_uuids, backend_state)
+
+    def filter_all(self, filter_obj_list, filter_properties):
+        """Yield objects that pass the filter.
+
+        Can be overridden in a subclass, if you need to base filtering
+        decisions on all objects.  Otherwise, one can just override
+        _filter_one() to filter a single object.
+
+        We override this so we can build a datastore -> backend fqdn lookup
+        for all the datastores.  The filter will use it to test if the
+        volume's backend fqdn matches the passed in pool's fqdn.
+        """
+        # reset the lookup for this request.  It can change between requests.
+        self.backend_fqdn_lookup = {}
+
+        # First we need to get all of the pools and populate a lookup for
+        # the datastore backend fqdn custom attributes, if any.
+        scheduler_hints = filter_properties.get('scheduler_hints') or {}
+        affinity_uuids = scheduler_hints.get(self.hint_key, [])
+        # filter_obj_list is a generator, so we need to copy
+        # each entry for looping again.
+        obj_list = []
+        if affinity_uuids:
+            # Build the backend_fqdn_lookup, since we will need it this request
+            objs = list(filter_obj_list)
+            for obj in objs:
```

Comment:
Why are you not simply iterating over `filter_obj_list`? In line 102, you set `obj_list = filter_obj_list` and iterate in 104 over it, so it should be possible, no?